src/pudl_archiver/archivers/classes.py

```diff
@@ -245,7 +249,86 @@ def _check_missing_files(
             description="Check for files from previous version of archive that would be deleted by the new version",
             success=len(missing_files) == 0,
             notes=notes,
-            required_for_run_success=self.check_missing_files,
+            required_for_run_success=self.fail_on_missing_files,
+        )
+
+    def _check_file_size(
+        self,
+        baseline_datapackage: DataPackage | None,
+        new_datapackage: DataPackage,
+    ) -> validate.DatasetSpecificValidation:
+        """Check if any one file's size has changed by |>allowed_file_rel_diff|."""
+        baseline_resources = set()
+        too_changed_files = set()
+        notes = None
+
+        if baseline_datapackage is not None:
+            baseline_resources = {
+                resource.name: resource.bytes_
+                for resource in baseline_datapackage.resources
+            }
+
+        new_resources = {
+            resource.name: resource.bytes_ for resource in new_datapackage.resources
+        }
+
+        # Check to see that file size hasn't changed by more than |>allowed_file_rel_diff|
+        # for each dataset in the baseline datapackage
+        for resource_name in baseline_resources:
+            if resource_name in new_resources:
+                file_size_change = abs(
+                    (new_resources[resource_name] - baseline_resources[resource_name])
+                    / baseline_resources[resource_name]
+                )
+                if file_size_change > self.allowed_file_rel_diff:
+                    too_changed_files.update({resource_name: file_size_change})
+
+        if too_changed_files:  # If files are "too changed"
+            notes = [
+                f"The following files have absolute changes in file size >|{self.allowed_file_rel_diff:.0%}|: {too_changed_files}"
+            ]
+
+        return validate.DatasetSpecificValidation(
+            name="Individual file size test",
+            description=f"Check for files from previous version of archive that have changed in size by more than {self.allowed_file_rel_diff:.0%}.",
+            success=not bool(too_changed_files),  # If dictionary empty, test passes
+            notes=notes,
+            required_for_run_success=self.fail_on_file_size_change,
+        )
+
+    def _check_dataset_size(
+        self,
+        baseline_datapackage: DataPackage | None,
+        new_datapackage: DataPackage,
+    ) -> validate.DatasetSpecificValidation:
+        """Check if a dataset's overall size has changed by more than |>allowed_dataset_rel_diff|."""
+        notes = None
+        baseline_size = None
+
+        if baseline_datapackage is not None:
```

Comment:
non-blocking: Similarly, it might be nice to return early from the `baseline_datapackage is None` case here, trading a bit of verbosity for less logic/state to carry around.

Actually in this function we get more out of the early return, since we then guarantee that `baseline_size is not None` in the rest of the function body. And then that `if baseline_size is not None` check can go away.

Reply:
I've moved things around here to move duplicate logic without returning twice. Let me know what you think!